storage/operation/children.go

```diff
@@ -9,34 +10,105 @@ import (
-//     This method silently overrides existing data without any sanity checks whether data for the same key already exits.
-//     Note that the Flow protocol mandates that for a previously persisted key, the data is never changed to a different
-//     value. Changing data could cause the node to publish inconsistent data and to be slashed, or the protocol to be
-//     compromised as a whole. This method does not contain any safeguards to prevent such data corruption. The lock proof
-//     serves as a reminder that the CALLER is responsible to ensure that the DEDUPLICATION CHECK is done elsewhere
-//     ATOMICALLY with this write operation.
+//   - The caller must acquire the [storage.LockInsertBlock] and hold it until the database write has been committed.
 //
-// No errors are expected during normal operation.
-func UpsertBlockChildren(lctx lockctx.Proof, w storage.Writer, blockID flow.Identifier, childrenIDs flow.IdentifierList) error {
-	held := lctx.HoldsLock(storage.LockInsertBlock) || lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock)
-	if !held {
-		return fmt.Errorf("missing required lock: %s or %s", storage.LockInsertBlock, storage.LockInsertOrFinalizeClusterBlock)
+// Expected error returns during normal operations:
+//   - [storage.ErrAlreadyExists] if the blockID is already indexed as a child of the parent
+func IndexNewBlock(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockInsertBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockInsertBlock)
 	}
 
-	return UpsertByKey(w, MakePrefix(codeBlockChildren, blockID), childrenIDs)
+	return indexBlockByParent(rw, blockID, parentID)
+}
+
+// IndexNewClusterBlock indexes a new cluster block and updates the parent-child relationship in the block children index.
+// This function creates an empty children index for the new cluster block and adds the new block to the parent's children list.
+//
+// CAUTION:
+//   - The caller must acquire the [storage.LockInsertOrFinalizeClusterBlock] and hold it until the database write has been committed.
+//
+// Expected error returns during normal operations:
+//   - [storage.ErrAlreadyExists] if the blockID is already indexed as a child of the parent
+func IndexNewClusterBlock(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
+	if !lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock) {
+		return fmt.Errorf("missing required lock: %s", storage.LockInsertOrFinalizeClusterBlock)
+	}
+
+	return indexBlockByParent(rw, blockID, parentID)
+}
+
+func indexBlockByParent(rw storage.ReaderBatchWriter, blockID flow.Identifier, parentID flow.Identifier) error {
+	// Step 1: make sure the new block has no children yet
+	var nonExist flow.IdentifierList
+	err := RetrieveBlockChildren(rw.GlobalReader(), blockID, &nonExist)
+	if err != nil {
```

Comment:
No need to check storage.ErrNotFound here, because `RetrieveBlockChildren` no longer returns it.